common/src/main/java/org/cbioportal/cmo/pipelines/common/util/HttpClientWithTimeoutAndRetry.java

```diff
@@ -0,0 +1,171 @@
+import java.time.temporal.ChronoUnit;
+import org.apache.log4j.Logger;
+import org.springframework.context.annotation.Bean;
+import org.springframework.http.client.SimpleClientHttpRequestFactory;
+import org.springframework.http.HttpEntity;
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.HttpMethod;
+import org.springframework.http.MediaType;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.client.RestClientException;
+import org.springframework.web.client.RestClientResponseException;
+import org.springframework.web.client.RestTemplate;
+
+public class HttpClientWithTimeoutAndRetry {
+    private int initialTimeout;
+    private int maximumTimeout;
+    private Instant dropDeadInstant;
+    private boolean retryOnErroneousServerResponse;
+    private int numberOfRequestsAttempted;
+    private int lastTimeoutUsed;
+    private boolean reachedDropDeadInstant;
+    private RestClientException lastRestClientException;
+    private String lastResponseBodyStringAfterException;
+    private Logger log = Logger.getLogger(HttpClientWithTimeoutAndRetry.class);
+
+    public HttpClientWithTimeoutAndRetry() {
+        this.initialTimeout = 1000; // default to 1 second initial timeout
+        this.maximumTimeout = 60000; // default to 1 minute maximum timeout
+        this.dropDeadInstant = null; // results in a default drop dead time of 4 * maximumTimeout (4 minutes) from start of exchange()
+        this.retryOnErroneousServerResponse = false;
+        this.numberOfRequestsAttempted = 0;
+        this.lastTimeoutUsed = 0;
+        this.reachedDropDeadInstant = false;
+        this.lastRestClientException = null;
+    }
+
+    public HttpClientWithTimeoutAndRetry(int initialTimeout, int maximumTimeout, Instant dropDeadInstant, boolean retryOnErroneousServerResponse) {
+        this.initialTimeout = initialTimeout;
+        this.maximumTimeout = maximumTimeout;
+        this.dropDeadInstant = dropDeadInstant;
+        this.retryOnErroneousServerResponse = retryOnErroneousServerResponse;
+        this.numberOfRequestsAttempted = 0;
+        this.lastTimeoutUsed = 0;
+        this.reachedDropDeadInstant = false;
+        this.lastRestClientException = null;
+    }
+
+    public int getNumberOfRequestsAttempted() {
+        return numberOfRequestsAttempted;
+    }
+
+    public int getLastTimeoutUsed() {
+        return lastTimeoutUsed;
+    }
+
+    public boolean getReachedDropDeadInstant() {
+        return reachedDropDeadInstant;
+    }
+
+    public RestClientException getLastRestClientException() {
+        return lastRestClientException;
+    }
+
+    public String getLastResponseBodyStringAfterException() {
+        return lastResponseBodyStringAfterException;
+    }
+
+    public boolean exceptionCausedByTimeout(RestClientException e) {
+        return e.getRootCause() instanceof SocketTimeoutException && e.getRootCause().getMessage().toLowerCase().contains("timed out");
```

Comment:
I do not know the answer, but I did see that the message included the text "timed out" in the timeouts caused by the exceeding of what was specified in setConnectTimeout() and setReadTimeout(). Those are the cases where I wanted our system to continue looping/retrying without logging. I kind of doubt that there are other SocketTimeoutException objects (which do not say "timed out" in the message), but just in case there are, these would be a little unexpected, so they would in fact be logged (in logger.debug()). Either way, the retry loop continues. I guess the difference is that expected timeout exceptions are not logged.